unify dynamic shapes API namings 3 (guard_int, guard_int_seq)

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #155973

evaluate_static_shape -> guard_int
evaluate_static_shapes -> guard_int_seq

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
[DRAFT] unify dynamic shapes API 3

[ghstack-poisoned]

laithsakka committed:
Update on "[DRAFT] unify dynamic shapes API 3"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bobrenjc93 commented:
lol if this is a draft maybe you should make it so?

laithsakka committed:
Update on "[DRAFT] unify dynamic shapes API 3 (guard_int, guard_int_seq)"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[DRAFT] unify dynamic shapes API 3 (guard_int, guard_int_seq)"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bobrenjc93 commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2359,7 +2359,7 @@ def _maybe_evaluate_static_worker(
         # variables have to have their value range bounds adjusted as
         # well)
-        s = sympy.Symbol(f"evaluate_static_shape_{idx}", positive=True, integer=True)
+        s = sympy.Symbol(f"guard_int_{idx}", positive=True, integer=True)
```
where is the guard happening? it seems off to me that you are saying we are guarding, which is a non pure function, in a pure function?

laithsakka replied:
this change is wrong. a log message happened to match the function name

laithsakka commented on torch/_inductor/sizevars.py:
```diff
@@ -486,15 +486,24 @@ def evaluate_max(self, left: Expr, right: Expr) -> Expr:
+        if isinstance(expr, int):
+            return expr
+        val = self.size_hint(expr)
+        self.check_equals(expr, sympy.Integer(val))
```
torch.check does guard/add runtime assert. this is similar to to torch.check(a==b)
the guard is added here in the impl of check_equals
```
 def check(self, expr: Expr) -> None:
        """
        Use it when you already know that expr is true or should be true and want to
        ensure that guards/runtime assertions are in place to ensure this in compiled
        function. Unlike expect_true, this WILL raise an error if expr isn't actually true.
        check Note [expect_true].
        """
        assert self.expect_true(expr)

    def check_equals(self, left: Expr, right: Expr) -> Expr:
        """
        check(sympy.Eq(left, right)) after applying inv_precomputed_replacements replacements.
        Returns left after applying inv_precomputed_replacements.
        """
        if isinstance(left, Expr):
            left = sympy_subs(left, self.inv_precomputed_replacements)  # type: ignore[arg-type]
        if isinstance(right, Expr):
            right = sympy_subs(right, self.inv_precomputed_replacements)  # type: ignore[arg-type]

        self.check(sympy.Eq(left, right))
  ```

guard_equeas(x, y) would mean in my mind evaluate x==y and return the results (True, or False).
check_equals(x, y) you use it when you know x should be y. an error is thrown if thats not right
see the previous PR https://github.com/pytorch/pytorch/pull/155776

bobrenjc93 replied:
> torch.check does guard/add runtime assert

Wait let's be precise. torch.check only adds runtime asserts right? So to be consistent doesn't it make sense to name this check_is_int?

bobrenjc93 reviewed (approved):
discussed offline @laithsakka will follow up with a separate PR

laithsakka committed:
Update on "unify dynamic shapes API namings 3 (guard_int, guard_int_seq)"


evaluate_static_shape -> guard_int
evaluate_static_shapes -> guard_int_seq

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "unify dynamic shapes API namings 3 (guard_int, guard_int_seq)"


evaluate_static_shape -> guard_int
evaluate_static_shapes -> guard_int_seq

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "unify dynamic shapes API namings 3 (guard_int, guard_int_seq)"


evaluate_static_shape -> guard_int
evaluate_static_shapes -> guard_int_seq

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge
